code/src/OSFramework/Grid/AbstractGrid.ts

```diff
@@ -81,20 +81,84 @@ namespace OSFramework.Grid {
             return this._provider;
         }
 
-        public get features(): OSFramework.Feature.ExposedFeatures {
+        public get features(): Feature.ExposedFeatures {
             return this._features;
         }
+        private _autoGenCol(): void {
+            //let's auto generate the columns
+            if (this.dataSource.hasMetadata) {
+                //if we have meta information about the columns, let's NOT use wijmo generator
+                this.autoGenerate = false;
+                const generated = this._columnsGenerator.generate(
+                    this,
+                    this.dataSource.getMetadata(),
+                    this.config.allowEdit
+                );
+                const newColumns = this._checkForNewColumns();
+                // remove existing columns if new dataSource has different columns
+                if (this._columns.size > 0 && newColumns) {
+                    this._columns.forEach((p) => this.removeColumn(p.uniqueId));
+                    generated.forEach((p) => this.addColumn(p));
+                }
+                // generate new columns
+                if (this._columns.size === 0 && newColumns) {
+                    generated.forEach((p) => this.addColumn(p));
+                }
+            } else {
+                //if the grid is read-only, then we'll flatten the array and use wijmo generator
+                if (!this.config.allowEdit) {
+                    this.dataSource.flatten();
+                } else {
+                    //if the grid is marked as editable, and is to be auto generated, we do not support (because of the save)
+                    throw new Error(
+                        'You cannot use JSONSerialize and make the grid editable. Please use ArrangeData action for this scenario.'
+                    );
+                }
+            }
+        }
 
+        private _checkForNewColumns(): boolean {
+            const metadata = this.dataSource.getMetadata();
+            let hasColumns = false;
+            return Object.keys(metadata).some((source) => {
+                const columns = Array.from(this._columns.values()).map(
+                    (col) => col.config.binding
+                );
+                const newColumns = Object.keys(metadata[source]);
+                hasColumns = newColumns.every((column) => {
+                    return columns.indexOf(`${source}.${column}`) !== -1;
```

Comment:
will this work when we have sources like -> "Sample_Product.Name" and "Name"?

Reply:
No :(((